// ngx-modialog/plugins/js-native/esm2015/lib/modal.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Overlay, DROP_IN_TYPE, Modal as Modal_ } from 'ngx-modialog';
import { JSNativePresetBuilder } from './presets/js-native-preset';
export class Modal extends Modal_ {
    /**
     * @param {?} overlay
     */
    constructor(overlay) {
        super(overlay);
    }
    /**
     * @return {?}
     */
    alert() {
        return new JSNativePresetBuilder(this, DROP_IN_TYPE.alert);
    }
    /**
     * @return {?}
     */
    prompt() {
        return new JSNativePresetBuilder(this, DROP_IN_TYPE.prompt);
    }
    /**
     * @return {?}
     */
    confirm() {
        return new JSNativePresetBuilder(this, DROP_IN_TYPE.confirm);
    }
    /**
     * @param {?} dialogRef
     * @param {?} type
     * @return {?}
     */
    create(dialogRef, type) {
        return dialogRef;
    }
}
Modal.decorators = [
    { type: Injectable },
];
/** @nocollapse */
Modal.ctorParameters = () => [
    { type: Overlay, },
];
function Modal_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    Modal.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    Modal.ctorParameters;
}
