// ngx-modialog/esm2015/public_api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
export { privateKey, setAssignMethod, setAssignAlias, FluentAssignFactory, FluentAssign } from './lib/framework/fluent-assign';
export { extend, arrayUnion, PromiseCompleter } from './lib/framework/utils';
export { createComponent } from './lib/framework/createComponent';
export { DialogBailOutError } from './lib/models/errors';
export { DialogRef } from './lib/models/dialog-ref';
export { DROP_IN_TYPE, OverlayRenderer } from './lib/models/tokens';
export { Modal, DOMOverlayRenderer } from './lib/providers/index';
export { overlayConfigFactory, OverlayContext, OverlayContextBuilder } from './lib/models/overlay-context';
export { Overlay, ModalOverlay, OverlayDialogBoundary, OverlayTarget } from './lib/overlay/index';
export { DEFAULT_VALUES, ModalContext, ModalContextBuilder } from './lib/models/modal-context';
export { ModalOpenContext, ModalOpenContextBuilder } from './lib/models/modal-open-context';
export { BaseDynamicComponent, CSSBackdrop, CSSDialogContainer } from './lib/components/index';
export { ModalModule } from './lib/ngx-modialog.module';
